compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1762,30 +1762,34 @@ impl<'a> Parser<'a> {
             span,
             "macros that expand to items must be delimited with braces or followed by a semicolon",
         );
-        if self.unclosed_delims.is_empty() {
-            let DelimSpan { open, close } = match args {
-                MacArgs::Empty | MacArgs::Eq(..) => unreachable!(),
-                MacArgs::Delimited(dspan, ..) => *dspan,
-            };
-            err.multipart_suggestion(
-                "change the delimiters to curly braces",
-                vec![(open, "{".to_string()), (close, '}'.to_string())],
-                Applicability::MaybeIncorrect,
-            );
-        } else {
+        // FIXME: This will make us not emit the help even for declarative
+        // macros within the same crate (that we can fix), which is sad.
+        if !span.from_expansion() {
```

Comment:
What happens if you instead use the following? Is the FIXME still applicable then? I suspect it might not be.

```suggestion
        if !span.can_be_used_for_suggestions() {
```

Reply:
That doesn't work. `can_be_used_for_suggestions()` only checks for derive macros where the span they emit is different from the span of themselves. This will both provide incorrect suggestions for those macros and still fail as with the FIXME.

https://github.com/rust-lang/rust/blob/49c82f31a85f04a709810de4ccfb8ba765c1377b/compiler/rustc_span/src/lib.rs#L575-L583

Reply:
What about using `in_derive_expansion` and only gate on those, then?

Reply:
I've thought about that, but attribute macros should not suggest help either; and probably neither should declarative macros from separate crates.

Reply:
Could you then checks for `!matches!(kind, ExprKind::Macro(MacroKind::Derive | MacroKind::Attr, _))` and using `tcx.sess().source_map().lookup_char_pos(span.lo()).file` on both the suggestion span and the callsite span to see if they belong to the same file?

Also, an additional test for those cases so that we see when the suggestion should appear in the same file might be useful. That might be easy to do for a macro by example. You might need to expand the test to also have an "external crate" example to check for that case too.

Would you have time to get these things done in this PR?

Reply:
Soon. But should it be the same file or same crate?

Reply:
And what about function-like proc macros? How can I differentiate those?

Reply:
Ah, you're right. @petrochenkov what would the best way of checking that a span doesn't correspond to a proc-macro of any type nor a macro by example from a foreign crate, while parsing an item?

Reply:
@petrochenkov friendly ping, if you have advice re: ^

I'm ok with landing with `if !span.from_expansion() {`, at least for now. Would you mind rebasing @ChayimFriedman2?

Reply:
@estebank ^^^ Done.